### ISSUE TITLE
Lazy load modules based on variables in path

### ISSUE THREAD

**Issue by vajahath** [angular/angular#29747]
# 🚀 feature request

### Relevant Package

_I hope_ this feature request is for the @angular/router

### Description

I don't know if this is already there in Angular, but I couldn't find any in the docs (or internet).

I think it's nice to have an option to specify routes like this:

`app-routing.module.ts`
```ts
import { Routes } from '@angular/router';

const routes: Routes = [
  {
    path: 'theme-builder/:themeId',
    loadChildren:
      'src/app/theme-builders/{{themeId}}-builder.module#BuilderModule'
  }
]
```

So that the angular can load the corresponding module. For large applications, sometimes we may not be able to specify every possible `loadChildren` paths. Sometimes these variables in the path are coming from a database - and it may also have a corresponding `loadChildren`. But specifying all these combinations hard-coded in the app doesn't feel like a scalable solution.

For clarity, the above code sample can be a replacement for:
```ts
const routes: Routes = [
  {
    path: 'theme-builder/theme-1',
    loadChildren:
      'src/app/theme-builders/theme-1-builder.module#BuilderModule'
  },
  {
    path: 'theme-builder/theme-2',
    loadChildren:
      'src/app/theme-builders/theme-2-builder.module#BuilderModule'
  },
  ...
]
```
### Describe the solution you'd like

I'm aware of the AOT. I think I can specify these individual lazy loaded modules in the `angular.josn` to compile them ahead of time.

```
{
  "architect":{
    "build":{
      "options":{
        "lazyModules": [
          "src/app/theme-builders/theme-1-builder.module",
          "src/app/theme-builders/theme-2-builder.module"
          ...
          // or even better a glob pattern
        ]
      }
    }
  }
}
```
### Describe alternatives you've considered

I came across this nice [article](https://netbasal.com/the-need-for-speed-lazy-load-non-routable-modules-in-angular-30c8f1c33093). But it seems to be more complex and doesn't support in-module-routing (I think).

**Comment by dawidgarus** [comment]
@vajahath What's the difference between explicitly listing all lazy modules in angular.json and router configuration? Also, what's the point of having the paths provided from database if you still need to have those modules (the actual files) defined in your code, so they are kind of hardcoded anyways.

**Comment by vajahath** [comment]
@dawidgarus Thank you for asking 🙂.

It's because I worry about load times.
If I'm hard-coding all these countless modules into the app-routing, while the actual user only needs one or two, I'm transferring lots of waste routing information over the wires. I think this increases the load time. Hard-coding 10 or 20 modules may not add much, but 100s* and plus does. 

But if we were able to specify a pattern in the app-routing like:
```
{
    path: 'theme-builder/:themeId',
    loadChildren:
      'src/app/theme-builders/{{themeId}}-builder.module#BuilderModule'
}
```
it's all there. We can compile individual modules in the back and serve them when asked without storing lots of routing information in the app hard-coded.

In the first, we are storing everything in the app. In the second, we are storing it in the `angular.json` which will not be served with the app. 🤔 ryt?

**Comment by Airblader** [comment]
And you have hundreds of those modules that are so fundamentally different that it can't be abstracted into a parameterized module?

**Comment by vajahath** [comment]
@Airblader There are cases where parameterized modules can help. But may not be a good fit in some other cases.

Consider this case: an application that lets you configure themes.
* The app contains a lot of themes you can configure and each theme comes with **its own theme builder**.
* Each theme is visually and functionally different - so their builder modules are too.
* For one theme, the theme builder needs only 1 page. For some other theme, we may need routes to organize complex building steps.

In this case, using a parameterized module for theme-builder may not be a good fit because, we've to compress and put every business logic into this single module. Moreover, the module have to load all the unwanted dependencies, while the actual logic needs only one or two.

**Comment by Airblader** [comment]
Fair enough. In any case, I wonder if Ivy doesn't perhaps implicitly resolve this problem already (maybe with the addition of a custom route reuse strategy). Someone with better Ivy knowledge could maybe answer this.

**Comment by dawidgarus** [comment]
@vajahath In this case I would try something like this. Forget about loadChildren and create a resolver that would return ComponentFactory for specific component based on path parameter. You can do something like this in your resolver (not actual code):
```
resolve(route) {
  return this.getModuleNgFactory(router.params.id)
    .then(m => m.create(this.injector).instance);
}
getModuleNgFactory(param) {
  if (this.compiler) { // inject optional Compiler to detect JIT
    return import(`something/${param}-builder.module`);
      .then(m => m[`${param}BuilderModule`]);
      .then(m => this.compiler.compileModuleAsync(m));
  } else {
    return import(`something/${param}-builder.module.ngfactory`)
      .then(m => m[`${param}BuilderModuleNgFactory`]);
  }
}
```
Module instance can have method to create the component to be displayed using ComponentFactory.

**Comment by dawidgarus** [comment]
@vajahath Also, if you worry for load times, test whether you should before you make your life deliberately harder. Gz works pretty good when there is a lot of repeating text. In my simple example: 1 route: 103 bytes after gz, 100 routes: 996 bytes, 1000 routes: 6786 bytes.

**Comment by vajahath** [comment]
Good.
I don't know, but don't we need the Angular compiler loaded into our app for that to work?
Maybe your code works (and that's great :star: ) but specifying all these like the following (or in a similar way)

```
{
    path: 'theme-builder/:themeId',
    loadChildren:
      'src/app/theme-builders/{{themeId}}-builder.module#BuilderModule'
}
```
feels more intuitive, cleaner and abstracted (at least for me :baby_bottle: ).

I'm happily leaving this to the Angular Team's choice to decide whether this feature request is good or not. May be I don't have the deeper knowledge to comment on this. I just pointed out an opportunity, where as a beginner, I struggled.
